File: day:6/isValid.js

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const VALID_CHARACTERS = ['(', ')', '{', '}', '[', ']'];
    const MAPPING = {
        '(': ')',
        '{': '}',
        '[': ']'
    }
    
    // base case(s)
    if(s.length === 0) return true;
    if(s.length === 1) return false;
    
    const stack = [];
    for(let x of s) {
        if(!VALID_CHARACTERS.includes(x)) return false;
        if(Object.keys(MAPPING).includes(x)) {
            stack.push(x);
        }
        else if(Object.values(MAPPING).includes(x)) {
            let left = stack.pop();
            if(MAPPING[left] !== x) {
                return false;
            } 
        }
    }
    if(stack.length === 0) {
        return true;
    }
    else {
        return false;
    }
    
};

File: day:6/isValid.java
class Solution {
    public boolean isValid(String s) {
        String VALID_CHARACTERS = "({[]})";
        Map<String, String> MAPPING = new HashMap<String, String>();
        MAPPING.put("(", ")");
        MAPPING.put("{", "}");
        MAPPING.put("[", "]");

        // base case(s)
        if(s.length() == 0) return true;
        if(s.length() == 1) return false;
    
        List<String> stack = new ArrayList<>();

        for(int i = 0; i < s.length(); i++) {
            String x = s.substring(i, i + 1);
            if(!VALID_CHARACTERS.contains(x)) return false;
            if(MAPPING.containsKey(x)) {
                stack.add(x);
            }
            else {
                if(stack.size() == 0) return  false;
                String left = stack.get(stack.size() - 1);
                if(!MAPPING.get(left).equals(x)) {
                    return false;
                }
                stack.remove(stack.size() - 1);
            }
        }
        return stack.size() == 0;
    }
}
